File: rust/src/protocol/v763/packets/play/clientbound/tab_complete.rs
//! Packet `tab_complete` (play/clientbound). Hand-written.

use crate::codec::uuid_codec::{self as uuid_c, Uuid};
use crate::codec::{
    bitset, chat_component, identifier, nbt, position, slot, string_codec as string, varint,
    varlong, BytesReader, BytesWriter, Reader, Writer,
};
use crate::errors::ProtocolError;
use crate::protocol::v763::states::ConnectionState;
use crate::protocol::v763::ClientboundPacket;

pub const PACKET_ID: i32 = 0x0F;

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct TabCompleteMatch {
    pub r#match: String,
    pub tooltip: Option<String>,
}

#[derive(Debug, Clone, PartialEq, Eq, Default)]
pub struct TabComplete {
    pub transaction_id: i32,
    pub start: i32,
    pub length: i32,
    pub matches: Vec<TabCompleteMatch>,
}

impl TabComplete {
    pub fn decode(reader: &mut BytesReader<'_>) -> Result<Self, ProtocolError> {
        let transaction_id = varint::read(reader)?;
        let start = varint::read(reader)?;
        let length = varint::read(reader)?;
        let n = varint::read(reader)? as usize;
        let mut matches = Vec::with_capacity(n);
        for _ in 0..n {
            let m = string::read(reader)?;
            let p = reader.read_exact(1)?[0];
            let tooltip = match p {
                0 => None,
                1 => Some(string::read(reader)?),
                other => return Err(ProtocolError::DecodeError(format!("tooltip: {}", other))),
            };
            matches.push(TabCompleteMatch {
                r#match: m,
                tooltip,
            });
        }
        Ok(Self {
            transaction_id,
            start,
            length,
            matches,
        })
    }
}

impl ClientboundPacket for TabComplete {
    fn state(&self) -> ConnectionState {
        ConnectionState::Play
    }
    fn packet_id(&self) -> i32 {
        PACKET_ID
    }
    fn encode(&self, writer: &mut BytesWriter) -> Result<(), ProtocolError> {
        varint::write(self.transaction_id, writer)?;
        varint::write(self.start, writer)?;
        varint::write(self.length, writer)?;
        varint::write(self.matches.len() as i32, writer)?;
        for m in &self.matches {
            string::write(&m.r#match, writer)?;
            match &m.tooltip {
                None => writer.write_all(&[0])?,
                Some(t) => {
                    writer.write_all(&[1])?;
                    string::write(t, writer)?;
                }
            }
        }
        Ok(())
    }
}

File: python/minecraft_bot/protocol/v763/packets/play/clientbound/tab_complete.py
"""Packet `tab_complete` (play/clientbound, id 0x0F).

Server's reply to a serverbound `tab_complete` request: a list of
matches, each with an optional tooltip.
"""

from __future__ import annotations

from dataclasses import dataclass

from minecraft_bot.codec import Reader, Writer, string, varint
from minecraft_bot.errors import ValueOutOfRange

PACKET_ID = 0x0F


@dataclass(frozen=True, slots=True)
class TabCompleteMatch:
    match: str
    tooltip: str | None  # JSON chat component, may be None


@dataclass(frozen=True, slots=True)
class TabComplete:
    transaction_id: int
    start: int
    length: int
    matches: tuple[TabCompleteMatch, ...]


def decode(reader: Reader) -> TabComplete:
    tid = varint.read(reader)
    st = varint.read(reader)
    ln = varint.read(reader)
    n = varint.read(reader)
    matches: list[TabCompleteMatch] = []
    for _ in range(n):
        m = string.read(reader)
        present = reader.read(1)[0]
        if present == 1:
            tip: str | None = string.read(reader)
        elif present == 0:
            tip = None
        else:
            raise ValueOutOfRange("tab_complete.match.tooltip.present", present)
        matches.append(TabCompleteMatch(match=m, tooltip=tip))
    return TabComplete(transaction_id=tid, start=st, length=ln, matches=tuple(matches))


def encode(packet: TabComplete, writer: Writer) -> None:
    varint.write(packet.transaction_id, writer)
    varint.write(packet.start, writer)
    varint.write(packet.length, writer)
    varint.write(len(packet.matches), writer)
    for m in packet.matches:
        string.write(m.match, writer)
        if m.tooltip is None:
            writer.write(b"\x00")
        else:
            writer.write(b"\x01")
            string.write(m.tooltip, writer)
